src/vm.rs

```diff
@@ -484,12 +487,62 @@ pub(crate) fn run(
                         break 'fail;
                     }
                 }
-                Insn::Lit(ref val) => {
-                    let ix_end = ix + val.len();
-                    if !matches_literal(s, ix, ix_end, val) {
+                Insn::Assertion(assertion) => {
+                    if !match assertion {
+                        Assertion::StartText => look_matcher.is_start(s.as_bytes(), ix),
+                        Assertion::EndText => look_matcher.is_end(s.as_bytes(), ix),
+                        Assertion::StartLine { crlf: false } => {
+                            look_matcher.is_start_lf(s.as_bytes(), ix)
+                        }
+                        Assertion::StartLine { crlf: true } => {
+                            look_matcher.is_start_crlf(s.as_bytes(), ix)
+                        }
+                        Assertion::EndLine { crlf: false } => {
+                            look_matcher.is_end_lf(s.as_bytes(), ix)
+                        }
+                        Assertion::EndLine { crlf: true } => {
+                            look_matcher.is_end_crlf(s.as_bytes(), ix)
+                        }
+                        Assertion::LeftWordBoundary => look_matcher
+                            .is_word_start_unicode(s.as_bytes(), ix)
+                            .unwrap(),
+                        Assertion::RightWordBoundary => {
+                            look_matcher.is_word_end_unicode(s.as_bytes(), ix).unwrap()
+                        }
+                        Assertion::WordBoundary => {
+                            look_matcher.is_word_unicode(s.as_bytes(), ix).unwrap()
+                        }
+                        Assertion::NotWordBoundary => look_matcher
+                            .is_word_unicode_negate(s.as_bytes(), ix)
+                            .unwrap(),
+                    } {
                         break 'fail;
                     }
-                    ix = ix_end;
+                }
+                Insn::Lit { ref val, casei } => {
+                    let end = ix.saturating_add(val.len()).min(s.len());
+                    let sb = &s.as_bytes()[ix..end];
+                    if !casei {
+                        if sb != val.as_bytes() {
+                            break 'fail;
+                        }
+                    } else {
+                        if !eq_test_exact_size(sb.iter().copied(), val.bytes(), |a, b| {
+                            // Do we need a == b shortcut? Is to_ascii_lowercase() fast?
+                            a == b || a.to_ascii_lowercase() == b.to_ascii_lowercase()
+                        }) {
+                            if !s.get(ix..end).map_or(false, |s| {
+                                eq_test(
+                                    s.chars().flat_map(char::to_lowercase),
+                                    val.chars().flat_map(char::to_lowercase),
```

Comment:
Was this copied from somewhere? Case-insensitive matching is a big can of worms. I'm not sure we should try to do it in fancy-regex. I'd prefer that we keep delegating these to regex.

Reply:
I'd prefer not to include this in this PR. It would need a lot of tests and as I said before, I'd rather just delegate this to regex for correctness. If it's in a separate PR we can look at benchmarks too.